### PR TITLE
Fix for func_id rollover

### PR DESCRIPTION
**Pull request by swimmadude66**
Since func_id is read from `magic` as a 16-bit signed integer, the `callFunction` method was receiving it in two's complement. This meant that any func id over 32,767 wrapped to a negative number, and thus could not be found in the context's fnMap.

This PR handles this by re-typing `magic` to a `uint32_t`, allowing over 4B func ids to be generated before overflow. There is a TS-imposed limit before this, though, as the [fnMap itself only supports 2^24 keys](https://stackoverflow.com/a/54466812/2212650). I've added a test that all values up to that number work, and to expect a rangeError when the map is overfull. (This test takes a LONG time, so feel free to remove. just wanted a test case that failed with the current version, and passed with the `uint32_t` change)

This change addresses our issues, but in general, I'd like to request some way to clean up the fnMap. Even once a function is dispose or set to undefined, the fnMap retains it, and the fnIndex increases forever. I'm not sure of the best fix for this, but it means eventually even this fix will cause us to get invalid function calls.

### CODE REVIEW

**Comment by swimmadude66** [comment]
@justjake please let me know if there's any more info you need around this, or if there's any other work needed. I know we have an odd usecase that can even register > 32K functions on a context, but its a real blocker for us.

File: ts/context.ts [hunk 1]
```diff
@@ -412,6 +412,7 @@ export class QuickJSContext implements LowLevelJavascriptVm<QuickJSHandle>, Disp
    */
   newFunction(name: string, fn: VmFunctionImplementation<QuickJSHandle>): QuickJSHandle {
     const fnId = ++this.fnNextId
+    // If fnID > 16777218 (24-bit max int) the map will error with too many keys
```
Comment (justjake) [inline ts/context.ts:415]: If you want this to scale, you can use an array of maps or an array-of-arrays; use the high byte to pick the index of the array to store the map/sub-array, and the rest of the bytes as the key/index in the map/sub-array.

File: ts/context.ts [hunk 2]
```diff
@@ -799,7 +800,9 @@ export class QuickJSContext implements LowLevelJavascriptVm<QuickJSHandle>, Disp
 
       const fn = this.fnMap.get(fn_id)
       if (!fn) {
-        throw new Error(`QuickJSContext had no callback with id ${fn_id}`)
+        console.error('Could not find function', fn_id)
+        process.exit(1)
```
Comment (justjake) [inline ts/context.ts]: It’s not appropriate to process.exit here.

File: ts/generated/ffi.WASM_DEBUG_ASYNCIFY.ts [hunk 3]
```diff
@@ -161,7 +161,7 @@ export class QuickJSAsyncFFI {
   QTS_BuildIsAsyncify: () => number =
     this.module.cwrap("QTS_BuildIsAsyncify", "number", [])
 
-  QTS_NewFunction: (ctx: JSContextPointer, func_id: number, name: string) => JSValuePointer =
+  QTS_NewFunction: (ctx: JSContextPointer, func_id: uint32_t, name: string) => JSValuePointer =
```
Comment (justjake) [inline ts/generated/ffi.WASM_DEBUG_ASYNCIFY.ts:164]: is uint32_t valid typescript?

File: ts/generated/ffi.WASM_RELEASE_ASYNCIFY.ts [hunk 4]
```diff
@@ -161,7 +161,7 @@ export class QuickJSAsyncFFI {
   QTS_BuildIsAsyncify: () => number =
     this.module.cwrap("QTS_BuildIsAsyncify", "number", [])
 
-  QTS_NewFunction: (ctx: JSContextPointer, func_id: number, name: string) => JSValuePointer =
+  QTS_NewFunction: (ctx: JSContextPointer, func_id: uint32_t, name: string) => JSValuePointer =
```
Comment (swimmadude66) [inline ts/generated/ffi.WASM_RELEASE_ASYNCIFY.ts:164]: looks like the generator is not properly mapping this to `number`. @justjake can you direct me to where that needs to be handled?

Comment (justjake) [reply ts/generated/ffi.WASM_RELEASE_ASYNCIFY.ts:164]: You can search the generator file for `int` and it should turn up.

**Review by justjake — COMMENTED** [review]
Do you need 2^24 functions registered at once, or can you make do with 2^16 functions if we can re-use function IDs after they’re dereferenced?

i would like to avoid modifying the QuickJS sources as much as possible. If the real fix here is re-using function IDs, I would prefer to do that than merge the current approach. 

As is, it seems like a lot of the issue could be solved by starting the function ID at the minimum int16_t, or by casting from signed to unsigned in our C interface code without modifying the interpreter.

File: ts/context.ts (hunk 1 above)
Comment (swimmadude66) [reply ts/context.ts:415]: That's a good idea. It still has an issue once the ids > Max 32-bit int, but that is quite a bit more space

File: ts/context.ts (hunk 2 above)
Comment (swimmadude66) [reply ts/context.ts]: yes sorry, this was from a debugging pass. its been removed

File: ts/generated/ffi.WASM_DEBUG_ASYNCIFY.ts (hunk 3 above)
Comment (swimmadude66) [reply ts/generated/ffi.WASM_DEBUG_ASYNCIFY.ts:164]: no, this needs to be mapped in the generator. Working on that part now

**Comment by swimmadude66** [comment]
> Do you need 2^24 functions registered at once, or can you make do with 2^16 functions if we can re-use function IDs after they’re dereferenced?
> 
> i would like to avoid modifying the QuickJS sources as much as possible. If the real fix here is re-using function IDs, I would prefer to do that than merge the current approach.
> 
> As is, it seems like a lot of the issue could be solved by starting the function ID at the minimum int16_t, or by casting from signed to unsigned in our C interface code without modifying the interpreter.

If function Ids could be re-used, I think that addresses most of our needs. 2^16 would be an improvement over the current situation, but more is definitely better. I agree that the minimum solution is treating magic as an unsigned int (16, 32, w/e) because the current issue is that `newFunction` registers as unsigned, but `callFunction` calls in as a signed int, so the ids do not match

File: ts/generated/ffi.WASM_RELEASE_ASYNCIFY.ts (hunk 4 above)
Comment (swimmadude66) [reply ts/generated/ffi.WASM_RELEASE_ASYNCIFY.ts]: found, fixed.

**Comment by justjake** [comment]
It’s possible to define a custom class from C code, and define a “finalizer” that’s called when an instance of the class is freed by the GC. I suggest investigating if you can define a custom class that behaves identically to a Function. Then in the finalizar, you can do some callback to “free” the function ID.

**Comment by swimmadude66** [comment]
> It’s possible to define a custom class from C code, and define a “finalizer” that’s called when an instance of the class is freed by the GC. I suggest investigating if you can define a custom class that behaves identically to a Function. Then in the finalizar, you can do some callback to “free” the function ID.

We actually are using FinalizationRegistry to do something similar, but it seems as though GC is not reliably firing. Even still, it appears functionIDs are purely incremented, so we'd still need some way to assign the next unused function ID. I think increasing the bit space for functions from approx 2^15 to 2^32 combined with some optimizations on our side will have us sorted for quite a bit

**Comment by swimmadude66** [comment]
Commenting to add that I was apparently incorrect before. Upon removing the changes from quickJS, we are still getting the conversion to negative numbers. I'm not sure exactly where that casting needs to be done, but its apparently missing here

**Comment by justjake** [comment]
It may be most simple to start with an initial ID of the minimum int16. Then no casting necessary?

File: ts/context.ts (hunk 1 above)
Comment (swimmadude66) [reply ts/context.ts]: I've implemented this as a map of maps with some protected helper functions for maintenance. Seems to work for the existing cases

**Comment by swimmadude66** [comment]
> It may be most simple to start with an initial ID of the minimum int16. Then no casting necessary?

That feels somewhat cloogier than simply using unsigned integers, no?

**Comment by justjake** [comment]
With regard to re-allocating function IDs, how do you know when a function is no longer needed by the guest? I was suggesting using finalizers inside the guest. Are you using FinalizationRegistry on the host, or inside the guest? In the case of the guest, there may be an API to explicitly run a GC cycle. In any case, you could push the no-longer-in-use function ID onto a “free-list”, and have the “get new function ID” operation prefer to pop from the free-list instead of bumping the counter.

**Comment by swimmadude66** [comment]
> With regard to re-allocating function IDs, how do you know when a function is no longer needed by the guest? I was suggesting using finalizers inside the guest. Are you using FinalizationRegistry on the host, or inside the guest? In the case of the guest, there may be an API to explicitly run a GC cycle. In any case, you could push the no-longer-in-use function ID onto a “free-list”, and have the “get new function ID” operation prefer to pop from the free-list instead of bumping the counter.

we are using FinalizationRegistry in the JS-land, and registering proxies to quickjs handles, using the same logic as the `Scope`. 

I will look into the reused function ids as a fix for the future, but I think that is substantially more complex (determining when is a function ready to be freed) than handling this negative rollover at the moment. I'm looking now to compare the effort of casting vs starting functionID at minimum signed int16

**Comment by justjake** [comment]
> > It may be most simple to start with an initial ID of the minimum int16. Then no casting necessary?
> 
> That feels somewhat cloogier than simply using unsigned integers, no?

Diverging from upstream is a bigger cludge i think.

**Comment by swimmadude66** [comment]
> > > It may be most simple to start with an initial ID of the minimum int16. Then no casting necessary?
> > 
> > 
> > That feels somewhat cloogier than simply using unsigned integers, no?
> 
> Diverging from upstream is a bigger cludge i think.

Fair enough. I'll make the min starting point change here, and look into a PR upstream to re-type `magic` to an unsigned int

**Comment by swimmadude66** [comment]
I've simplified the changeset and opened an issue with quickjs. Please let me know if you'd like me to make any other changes before this can land. Definitely looking forward to using the full address space for functions

**Review by justjake — APPROVED** [review]
Looks good to me!

**Comment by justjake** [comment]
By the way, to work around the issue entirely, you could use a mapping table in host user space backed by a UUID. Instead of binding 2^N functions via newFunction, just make one called dispatch, and then call dispatch.bind(uuid) *inside the VM*.

**Comment by justjake** [comment]
@swimmadude66 it looks like the tests failed

**Comment by swimmadude66** [comment]
> @swimmadude66 it looks like the tests failed. Can you get them to pass?

I've pushed a "fix" which skips the massive max funcID test in debug mode, as it was timing out (and takes ~5min locally). I can re-add it with an absurd timeout if you'd like, but I think its probably ok to skip this one in debug

**Comment by justjake** [comment]
Sounds good

**Comment by swimmadude66** [comment]
@justjake sorry, I missed a place for my fix. re-pushed, but the tests don't seem to re-run automatically

**Pull request merged**